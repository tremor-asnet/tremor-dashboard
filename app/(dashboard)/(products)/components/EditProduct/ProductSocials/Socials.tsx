"use client";

import { Controller, useFormContext } from "react-hook-form";

// Components
import { Title, Flex, Card } from "@tremor/react";

//Styles
import "@/styles/products.css";

const Socials = () => {
  const { control } = useFormContext();

  return (
    <Card className="dark:bg-dark-tremor-primary ring-0">
      <Title className="font-primary font-bold text-primary dark:text-dark-primary text-xl leading-snug capitalize">
        Socials
      </Title>
      <Flex className="flex-col items-start justify-start mt-8">
        <Controller
          control={control}
          render={({ field }) => (
            <input
              className="input-text"
              id="shoppify-handle"
              placeholder="Shoppify Handle"
              {...field}
            />
          )}
          name="shopifyUrl"
        />
        <Controller
          control={control}
          render={({ field }) => (
            <input
              className="input-text"
              id="facebook-account"
              placeholder="Facebook Account"
              {...field}
            />
          )}
          name="facebookUrl"
        />
        <Controller
          control={control}
          render={({ field }) => (
            <input
              className="input-text"
              id="instagram-account"
              placeholder="Instagram Account"
              {...field}
            />
          )}
          name="instagramUrl"
        />
      </Flex>
    </Card>
  );
};

export default Socials;
