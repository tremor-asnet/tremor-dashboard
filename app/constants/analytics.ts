/* Constants for Analytics Page */
export const SALE_STATISTICAL_TYPE = {
  REVENUE: "Revenue",
};
