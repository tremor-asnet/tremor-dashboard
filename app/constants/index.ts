export * from "./routes";
export * from "./regex";
export * from "./messages";
export * from "./profile";
export * from "./commons";
export * from "./flagSrc";
export * from "./images";
export * from "./sections";
export * from "./images";
export * from "./products";
export * from "./colors";
export * from "./units";
export * from "./sales";
export * from "./status";
export * from "./saleStatistical";
export * from "./analytics";
export * from "./breakpoints";
export * from "./billing";
export * from "./pincode";
export * from "./api";
