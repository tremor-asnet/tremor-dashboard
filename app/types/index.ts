export * from "./user";
export * from "./breadcrumb";
export * from "./tab";
export * from "./project";
export * from "./analytics";
export * from "./sales";
export * from "./profile";
export * from "./table-list";
export * from "./orders";
