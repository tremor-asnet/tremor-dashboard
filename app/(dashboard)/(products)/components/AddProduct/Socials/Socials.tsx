// Libs
import { Controller, FieldErrors } from "react-hook-form";

// Components
import InputField from "@/components/common/CustomField/InputField/InputField";

// Constants
import { MESSAGES_ERROR, REGEX } from "@/constants";

// Types
import { NewSocial } from "@/types";

interface SocialsProps {
  control: any;
  errors: FieldErrors<NewSocial>;
}

const Socials = ({ control, errors }: SocialsProps) => {
  const facebookErrorMsg = errors.facebookUrl?.message;
  const instagramErrorMsg = errors.instagramUrl?.message;
  const shopifyErrorMsg = errors.shopifyUrl?.message;

  return (
    <div className="flex flex-col gap-6">
      <Controller
        name="shopifyUrl"
        control={control}
        rules={{
          required: MESSAGES_ERROR.URL_REQUIRED,
          pattern: {
            value: REGEX.URL,
            message: MESSAGES_ERROR.URL_REQUIRED,
          },
        }}
        render={({ field }) => (
          <>
            <InputField
              label="Shopify Handle"
              placeholder=""
              type="text"
              {...field}
            />
            <p className="text-[11px] xs:text-xs text-red-500">
              {shopifyErrorMsg}
            </p>
          </>
        )}
      />

      <Controller
        name="facebookUrl"
        control={control}
        rules={{
          required: MESSAGES_ERROR.URL_REQUIRED,
          pattern: {
            value: REGEX.URL,
            message: MESSAGES_ERROR.URL_REQUIRED,
          },
        }}
        render={({ field }) => (
          <>
            <InputField
              label="Facebook Account"
              placeholder=""
              type="text"
              {...field}
            />
            <p className="text-[11px] xs:text-xs text-red-500">
              {facebookErrorMsg}
            </p>
          </>
        )}
      />

      <Controller
        name="instagramUrl"
        control={control}
        rules={{
          required: MESSAGES_ERROR.URL_REQUIRED,
          pattern: {
            value: REGEX.URL,
            message: MESSAGES_ERROR.URL_REQUIRED,
          },
        }}
        render={({ field }) => (
          <>
            <InputField
              label="Instagram Account"
              placeholder=""
              type="text"
              {...field}
            />
            <p className="text-[11px] xs:text-xs text-red-500">
              {instagramErrorMsg}
            </p>
          </>
        )}
      />
    </div>
  );
};

export default Socials;
