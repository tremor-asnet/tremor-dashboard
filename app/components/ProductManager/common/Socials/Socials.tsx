// Libs
import { Control, Controller } from "react-hook-form";

// Components
import InputField from "@/components/common/CustomField/InputField/InputField";

// Types
import { NewSocial } from "@/types";

interface SocialsProps {
  control: Control<NewSocial>;
}

const Socials = ({ control }: SocialsProps) => (
  <div className="flex flex-col gap-6">
    <Controller
      name="shopifyUrl"
      control={control}
      render={({ field }) => (
        <InputField
          label="Shopify Handle"
          placeholder=""
          type="text"
          {...field}
        />
      )}
    />

    <Controller
      name="facebookUrl"
      control={control}
      render={({ field }) => (
        <InputField
          label="Facebook Account"
          placeholder=""
          type="text"
          {...field}
        />
      )}
    />

    <Controller
      name="instagramUrl"
      control={control}
      render={({ field }) => (
        <InputField
          label="Instagram Account"
          placeholder=""
          type="text"
          {...field}
        />
      )}
    />
  </div>
);

export default Socials;
