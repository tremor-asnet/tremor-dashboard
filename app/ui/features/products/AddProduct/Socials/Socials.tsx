// Libs
import { Controller, FieldErrors } from "react-hook-form";

// Components
import { InputField } from "@/ui/components";
import { Text } from "@tremor/react";

// Constants
import { MESSAGES_ERROR, REGEX } from "@/constants";

// Types
import { NewSocial } from "@/types";

interface SocialsProps {
  control: any;
  errors: FieldErrors<NewSocial>;
}

const Socials = ({ control, errors }: SocialsProps) => {
  const facebookErrorMsg = errors.facebookUrl?.message;
  const instagramErrorMsg = errors.instagramUrl?.message;
  const shopifyErrorMsg = errors.shopifyUrl?.message;

  return (
    <>
      <Controller
        name="shopifyUrl"
        control={control}
        rules={{
          required: MESSAGES_ERROR.URL_REQUIRED,
          pattern: {
            value: REGEX.URL,
            message: MESSAGES_ERROR.URL_REQUIRED,
          },
        }}
        render={({ field }) => (
          <div className="mb-3">
            <InputField
              id="add-product-shopify"
              label="Shopify Handle"
              type="text"
              {...field}
            />
            <Text className="text-xs text-red-500 dark:text-red-500 mt-1">
              {shopifyErrorMsg}
            </Text>
          </div>
        )}
      />

      <Controller
        name="facebookUrl"
        control={control}
        rules={{
          required: MESSAGES_ERROR.URL_REQUIRED,
          pattern: {
            value: REGEX.URL,
            message: MESSAGES_ERROR.URL_REQUIRED,
          },
        }}
        render={({ field }) => (
          <div className="mb-3">
            <InputField
              id="add-product-fb"
              label="Facebook Account"
              type="text"
              {...field}
            />
            <Text className="text-xs text-red-500 dark:text-red-500 mt-1">
              {facebookErrorMsg}
            </Text>
          </div>
        )}
      />

      <Controller
        name="instagramUrl"
        control={control}
        rules={{
          required: MESSAGES_ERROR.URL_REQUIRED,
          pattern: {
            value: REGEX.URL,
            message: MESSAGES_ERROR.URL_REQUIRED,
          },
        }}
        render={({ field }) => (
          <div className="mb-3">
            <InputField
              id="add-product-ig"
              label="Instagram Account"
              type="text"
              {...field}
            />
            <Text className="text-xs text-red-500 dark:text-red-500 mt-1">
              {instagramErrorMsg}
            </Text>
          </div>
        )}
      />
    </>
  );
};

export default Socials;
