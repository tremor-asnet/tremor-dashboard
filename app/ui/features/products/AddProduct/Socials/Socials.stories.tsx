import type { Meta, StoryObj } from "@storybook/react";

// Libs
import { useForm } from "react-hook-form";

// Components
import Socials from "./Socials";

// Types
import { NewSocial } from "@/types";

const meta = {
  title: "Components/Products/AddProduct/Socials",
  component: Socials,
  tags: ["autodocs"],
  argTypes: {
    control: {
      description:
        "This is control from useForm of react-hook-form with type NewSocial(shopifyUrl: string, facebookUrl: string, instagramUrl: string)",
    },
    errors: {
      description:
        "This is errors from FieldErrors of react-hook-form with type NewSocial()",
    },
  },
} as Meta<typeof Socials>;

export default meta;

type Story = StoryObj<typeof meta>;

const SocialsFormProvider = () => {
  const { control } = useForm<NewSocial>({
    defaultValues: {
      shopifyUrl: "https://shoppify-url.com",
      facebookUrl: "https://facebook-url.com",
      instagramUrl: "https://instagra-url.com",
    },
  });

  return <Socials control={control} />;
};

export const Default: Story = {
  render: () => <SocialsFormProvider />,
};
