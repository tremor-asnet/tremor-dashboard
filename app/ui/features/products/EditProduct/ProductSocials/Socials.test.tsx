import { render } from "@testing-library/react";

// Constants
import { MESSAGES_ERROR } from "@/constants";

// Components
import Socials from "./Socials";
import { useForm } from "react-hook-form";

// Types
import { NewSocial } from "@/types";

const mockControllerNonError = (name: string, rule: Object) => ({
  field: {
    name: name,
    rules: rule,
  },
  formState: { errors: {} },
});

// Mock react-hook-form module
jest.mock("react-hook-form", () => ({
  ...jest.requireActual("react-hook-form"),
  // Mock the useFormContext hook
  useFormContext: jest.fn(() => ({
    formState: {
      errors: {},
    },
  })),
  // Mock the Controller component
  Controller: (props: any) => {
    return props.render(mockControllerNonError(props.name, props.rule));
  },
}));

describe("Socials Component", () => {
  it("Should match snapshot", () => {
    const { container } = render(<Socials />);
    expect(container).toMatchSnapshot();
  });

  it("Should show error message when all field has invalid URL", () => {
    // Override the mock implementation of Controller for this test case
    jest.requireMock("react-hook-form").Controller = (props: any) => {
      return props.render({
        formState: {
          field: {
            name: props.name,
            rules: props.rule,
          },
          errors: {
            shopifyUrl: { message: MESSAGES_ERROR.INVALID_URL },
            facebookUrl: { message: MESSAGES_ERROR.INVALID_URL },
            instagramUrl: { message: MESSAGES_ERROR.INVALID_URL },
          },
        },
      });
    };

    const { getAllByText } = render(<Socials />);
    expect(getAllByText(MESSAGES_ERROR.INVALID_URL)).toHaveLength(3);
  });

  it("Should not show error messages for valid Shopify url", () => {
    const validShopifyURL = "https://valid.shopify.url";
    // Override the mock implementation of Controller for this test case
    jest.requireMock("react-hook-form").Controller = (props: any) => {
      if (props.name === "shopifyUrl")
        return props.render({
          field: {
            name: props.name,
            rules: props.rule,
            onChange: jest.fn(),
            value: [validShopifyURL],
          },
          formState: {
            errors: {},
          },
        });
      return props.render(mockControllerNonError(props.name, props.rule));
    };

    const { getAllByDisplayValue, queryByText } = render(<Socials />);
    expect(getAllByDisplayValue(validShopifyURL)).toHaveLength(1);
    expect(queryByText(MESSAGES_ERROR.INVALID_URL)).toBeNull();
  });

  it("Should not show error messages for valid Facebook url", () => {
    const validFacebookURL = "https://valid.facebook.url";
    // Override the mock implementation of Controller for this test case
    jest.requireMock("react-hook-form").Controller = (props: any) => {
      if (props.name === "facebookUrl")
        return props.render({
          field: {
            name: props.name,
            rules: props.rule,
            onChange: jest.fn(),
            value: [validFacebookURL],
          },
          formState: {
            errors: {},
          },
        });
      return props.render(mockControllerNonError(props.name, props.rule));
    };

    const { getAllByDisplayValue, queryByText } = render(<Socials />);
    expect(getAllByDisplayValue(validFacebookURL)).toHaveLength(1);
    expect(queryByText(MESSAGES_ERROR.INVALID_URL)).toBeNull();
  });

  it("Should not show error messages for valid Instagram url", () => {
    const validInstagramURL = "https://valid.instagram.url";
    // Override the mock implementation of Controller for this test case
    jest.requireMock("react-hook-form").Controller = (props: any) => {
      if (props.name === "instagramUrl")
        return props.render({
          field: {
            name: props.name,
            rules: props.rule,
            onChange: jest.fn(),
            value: [validInstagramURL],
          },
          formState: {
            errors: {},
          },
        });
      return props.render(mockControllerNonError(props.name, props.rule));
    };

    const { getAllByDisplayValue, queryByText } = render(<Socials />);
    expect(getAllByDisplayValue(validInstagramURL)).toHaveLength(1);
    expect(queryByText(MESSAGES_ERROR.INVALID_URL)).toBeNull();
  });

  it("Should show error messages for invalid Shopify url", () => {
    const invalidShopifyURL = "invalid.shopify.url";
    // Override the mock implementation of Controller for this test case
    jest.requireMock("react-hook-form").Controller = (props: any) => {
      if (props.name === "shopifyUrl")
        return props.render({
          field: {
            name: props.name,
            rules: props.rule,
            onChange: jest.fn(),
            value: [invalidShopifyURL],
          },
          formState: {
            errors: {
              shopifyUrl: { message: MESSAGES_ERROR.INVALID_URL },
            },
          },
        });
      return props.render(mockControllerNonError(props.name, props.rule));
    };

    const { getAllByDisplayValue, getAllByText } = render(<Socials />);
    const inputShopifyUrl = getAllByDisplayValue(invalidShopifyURL);
    expect(inputShopifyUrl).toHaveLength(1);
    expect(getAllByText(MESSAGES_ERROR.INVALID_URL)).toHaveLength(1);

    const parents = inputShopifyUrl.map(input => input.parentElement);
    const errorMessage = parents.map(parent => parent?.querySelector("p"));

    errorMessage.forEach(p => {
      expect(p?.textContent).toBe(MESSAGES_ERROR.INVALID_URL);
    });
  });

  it("Should show error messages for invalid Facebook url", () => {
    const invalidFacebookURL = "invalid.facebook.url";
    // Override the mock implementation of Controller for this test case
    jest.requireMock("react-hook-form").Controller = (props: any) => {
      if (props.name === "facebookUrl")
        return props.render({
          field: {
            name: props.name,
            rules: props.rule,
            onChange: jest.fn(),
            value: [invalidFacebookURL],
          },
          formState: {
            errors: {
              facebookUrl: { message: MESSAGES_ERROR.INVALID_URL },
            },
          },
        });
      return props.render(mockControllerNonError(props.name, props.rule));
    };

    const { getAllByDisplayValue, getAllByText } = render(<Socials />);
    const inputShopifyUrl = getAllByDisplayValue(invalidFacebookURL);
    expect(inputShopifyUrl).toHaveLength(1);
    expect(getAllByText(MESSAGES_ERROR.INVALID_URL)).toHaveLength(1);

    const parents = inputShopifyUrl.map(input => input.parentElement);
    const errorMessage = parents.map(parent => parent?.querySelector("p"));

    errorMessage.forEach(p => {
      expect(p?.textContent).toBe(MESSAGES_ERROR.INVALID_URL);
    });
  });

  it("Should show error messages for invalid Instagram url", () => {
    const invalidInstagramURL = "invalid.instagram.url";
    // Override the mock implementation of Controller for this test case
    jest.requireMock("react-hook-form").Controller = (props: any) => {
      if (props.name === "instagramUrl")
        return props.render({
          field: {
            name: props.name,
            rules: props.rule,
            onChange: jest.fn(),
            value: [invalidInstagramURL],
          },
          formState: {
            errors: {
              instagramUrl: { message: MESSAGES_ERROR.INVALID_URL },
            },
          },
        });
      return props.render(mockControllerNonError(props.name, props.rule));
    };

    const { getAllByDisplayValue, getAllByText } = render(<Socials />);
    const inputShopifyUrl = getAllByDisplayValue(invalidInstagramURL);
    expect(inputShopifyUrl).toHaveLength(1);
    expect(getAllByText(MESSAGES_ERROR.INVALID_URL)).toHaveLength(1);

    const parents = inputShopifyUrl.map(input => input.parentElement);
    const errorMessage = parents.map(parent => parent?.querySelector("p"));

    errorMessage.forEach(p => {
      expect(p?.textContent).toBe(MESSAGES_ERROR.INVALID_URL);
    });
  });
});
