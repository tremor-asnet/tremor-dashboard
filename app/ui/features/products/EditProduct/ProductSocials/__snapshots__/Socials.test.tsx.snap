// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Socials Component Should match snapshot 1`] = `
<div>
  <div
    class="tremor-Card-root relative w-full text-left rounded-tremor-default p-6 bg-tremor-background ring-tremor-ring shadow-tremor-card dark:ring-dark-tremor-ring dark:shadow-dark-tremor-card border-tremor-brand dark:border-dark-tremor-brand dark:bg-dark-tremor-primary ring-0"
  >
    <p
      class="text-tremor-title font-primary font-bold text-primary dark:text-dark-primary text-xl leading-snug capitalize"
    >
      Socials
    </p>
    <div
      class="tremor-Flex-root flex w-full flex-col justify-start items-start mt-4 gap-6"
    >
      <div
        class="w-full mb-4"
      >
        <div
          class="w-full"
        >
          <label
            class="text-secondary text-sm dark:text-lighter"
            for="edit-shopify"
          >
            Shopify Handle
          </label>
          <input
            class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
            id="edit-shopify"
            name="shopifyUrl"
            placeholder=""
            type="text"
          />
        </div>
      </div>
      <div
        class="w-full mb-4"
      >
        <div
          class="w-full"
        >
          <label
            class="text-secondary text-sm dark:text-lighter"
            for="edit-fb"
          >
            Facebook Account
          </label>
          <input
            class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
            id="edit-fb"
            name="facebookUrl"
            placeholder=""
            type="text"
          />
        </div>
      </div>
      <div
        class="w-full mb-4"
      >
        <div
          class="w-full"
        >
          <label
            class="text-secondary text-sm dark:text-lighter"
            for="edit-ig"
          >
            Instagram Account
          </label>
          <input
            class="block py-2.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer"
            id="edit-ig"
            name="instagramUrl"
            placeholder=""
            type="text"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
